File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const N = maps.length;
    const M = maps[0].length;
    const dx = [-1, 1, 0, 0];
    const dy = [0, 0, -1, 1];
    
    let tovisit = [];
    tovisit.push([0,0]);
    
    while (tovisit.length !== 0) {
        const [cx, cy] = tovisit.shift();
        for(let i=0;i<4;i++) {
            const nx = cx + dx[i];
            const ny = cy + dy[i];
            
            if (0 <= nx && nx < N && 0 <= ny && ny < M && maps[nx][ny] == 1) {
                tovisit.push([nx, ny]);
                maps[nx][ny] = maps[cx][cy] + 1;
            }
        }
    }
    
    return maps[N-1][M-1] === 1? -1: maps[N-1][M-1];
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n, m = len(maps), len(maps[0])
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    def bfs(x, y):
        tovisit = deque([])
        tovisit.append((x, y))
        
        while tovisit:
            cx, cy = tovisit.popleft()
            
            for i in range(4):
                next_x = cx + dx[i]
                next_y = cy + dy[i]
                
                if 0 <= next_x < n and 0 <= next_y < m and maps[next_x][next_y] == 1:
                    tovisit.append((next_x, next_y))
                    maps[next_x][next_y] = maps[cx][cy]+1
        return maps[n-1][m-1] if maps[n-1][m-1] != 1 else -1

    return bfs(0,0)